erts/emulator/beam/erl_bif_info.c

```diff
@@ -2021,21 +2022,44 @@ current_function(Process *c_p, ErtsHeapFactory *hfact, Process* rp,
     }
 
     if (c_p == rp && !(flags & ERTS_PI_FLAG_REQUEST_FOR_OTHER)) {
-	FunctionInfo fi2;
-        BeamInstr* continuation_ptr;
+        BeamInstr* return_address;
+        FunctionInfo caller_fi;
+        Eterm *ptr;
 
-	/*
-	 * The current function is erlang:process_info/{1,2},
-	 * which is not the answer that the application want.
-         * We will use the continuation pointer stored at the
-         * top of the stack instead.
-	 */
-        continuation_ptr = (BeamInstr *) rp->stop[0];
-        erts_lookup_function_info(&fi2, continuation_ptr, full_info);
-	if (fi2.mfa) {
-	    fi = fi2;
-	    rp->current = fi2.mfa;
-	}
+        /*
+         * The current function is erlang:process_info/{1,2}, and we've
+         * historically returned the *calling* function in that case. We
+         * therefore use the continuation pointer stored at the top of the
+         * stack instead, which is safe since process_info is a "heavy" BIF
+         * that is only called through its export entry.
+         */
+
+        return_address = NULL;
+        ptr = STACK_TOP(rp);
+        ASSERT(is_CP(*ptr));
+
+        while (ptr < STACK_START(rp)) {
+            BeamInstr *cp = cp_val(*ptr);
+
+            if (*cp == BeamOpCodeAddr(op_return_trace)) {
```

Comment:
Third time I see this loop, could it be refactored into one function?

Reply:
Good point, I'll push a commit that uses the `printable_return_address` function instead.